functions/src/notifications.ts

```diff
@@ -92,6 +92,75 @@ export function createActivityNotification(token: string, activity_id: string):
     }
 }
 
+// Sends a message to a chat if a user leaves or joins it:
+export async function onActivityUsersChanged(change: Change<FirebaseFirestore.DocumentSnapshot>, context: EventContext) {
+    const activityId: string = change.after.id
+    const usersBefore = getUsersFromChange(change.before)
+    const usersAfter = getUsersFromChange(change.after)
+
+    if (usersBefore.length !== usersAfter.length) {
+        // Gets which user joined or left the activity. changedUser will be that person's UID.
+        const changedUser = getDifferentString(usersBefore, usersAfter)
+        if (! changedUser) return -1
+
+        // @ts-ignore because it doesn't like the mock
+        const userInfo = await Refs(database).user(changedUser).get()
+        if (userInfo.exists) {
+            const changedUserInfo = userInfo.data()
+            // Setup message based on whether they left or joined:
+            const msgBody = `${changedUserInfo.name} has ${usersBefore > usersAfter ? 'left' : 'joined'} your activity.`
+            
+            return exports.sendChatMessage(activityId, msgBody, changedUser, new Date())
+        }
+    }
+    return 0
+}
+
+// Sends a user left/joined message to the specified activity
+export function sendChatMessage(activityId: string, message: string, sender: string, date: Date) {
+    // @ts-ignore because it doesn't like the database mock:
+    const chatRef = Refs(database).chat(activityId)
+    return chatRef.add({
+        message,
+        sender,
+        date_sent: date,
+        type: 'user_join_or_leave'
+    })
+}
+
+// Gets the the array of activity members from the activity doc.
+export function getUsersFromChange(doc: FirebaseFirestore.DocumentSnapshot): string[] {
+    const data1 = doc.data()
+    const arr1: string[] = (data1 && data1.members && Array.isArray(data1.members)) ? data1.members : []
+    return arr1
+}
+
+// Returns the first string which is different in the array.
+// Returns null if the arrays are the same length.
+export function getDifferentString(before: string[], after: string[]): string | null {
+    // Get the strings in the same order:
+    before.sort()
+    after.sort()
+    // Someone has been removed:
+    if (before.length > after.length) {
```

Comment:
Do we think adding Lodash would be worthwhile?  We could just call `_.difference()` and have less code to maintain and test.

Reply:
I think you raise some good points - I think lodash would be good to add just for this.

Reply:
Agreed!  I bet it could come in handy later on too.